### DISCUSSION TITLE
How to properly unsubscribe from a topic (Streaming API)?

### DISCUSSION THREAD

**Discussion by mikunn** [jsforce/jsforce#1451] [Q&A]
I'm new to Salesforce and can't get my head around how to property unsubscribe from a topic. The following piece of code doesn't seem to never quit and I have to force the exit. However, it will exit property if I comment out lines marked with `//1` and `//2`.

```
const jsforce = require('jsforce');

let loginUrl = "...";
let username = "...";
let password = "...";

let conn = new jsforce.Connection({loginUrl: loginUrl});
let subscription;

conn.login(username, password, function() {
  console.log('logged in');
  subscription = conn.streaming.topic('foobar').subscribe(cb);  //1
  setTimeout(() => unsub(), 3000);
});

function cb(message) {
  console.log('message', message);
}

function unsub() {
  console.log('unsubbing');
  conn.streaming.topic('foobar').unsubscribe(cb);
  subscription.cancel();                                         //2
  conn.logout((err) => console.log('conn logout err', err));
}
```

The code outputs
```
logged in
unsubbing
conn logout err null
```
JSforce version: 1.8.0
Node version: 7.2.1

What am I doing wrong? Thanks for any help!

**Comment by jan-goebel-lendico** [comment]
@mikunn How did you resolve this?

**Comment by ahmetkocabiyik** [comment]
Same problem for me. I tried like @mikunn but no workaround. Someone can share an example to properly unsubscribe from a topic.

**Comment by Haitaar** [comment]
Hey everyone, same here. Has anyone found a way to properly unsubscribe from a topic?

**Comment by AllanOricil** [comment]
The unsubscribe method does not work.

**Comment by jsonpoindexter** [comment]
Has anyone has success with unsubscribing?

**Comment by jsonpoindexter** [comment]
Ok, so I did end up figuring out how to successfully unsubscribe. The issue seems to come down to callbacks within the `.subscribe()` method needing to resolve before you can call `.unsubscribe()`

For example this will not unsubscribe:
```
const s = conn.streaming.topic('foobar');
s.subscribe(cb);
s.unsubscribe(cb); // Does not work
```

but this will work:
```
const s = conn.streaming.topic('foobar');
s.subscribe(cb);
setTimeout(() => {
    s.unsubscribe() // Works!
}, 1000)
```

Seeing as how the `.subscribe()` method does not return a promise I don't see a way to actually know if  `subscribe()` has finished or not before calling `unsubscribe()`

In my case I also saved the subscription to later then call `.unsubscribe()` on. Using `conn.streaming.topic('foobar').unsubscribe(cb);` did not work for me.

FYI @mikunn  @jan-goebel-lendico  @ahmetkocabiyik  @Haitaar @AllanOricil

**Comment by AllanOricil** [comment]
@jsonpoindexter I just went to my old code and saw that I was calling the unsub with the click of a button and it was not working. I did not see any difference from your code and mine. Which jsforce version are you using?

**Comment by jsonpoindexter** [comment]
@AllanOricil  my jsforce version is `1.10.1` using  `Nodejs 14.17.1`.

It sounds like you are running in the browser?

**Comment by AllanOricil** [comment]
@jsonpoindexter Im not. I have a node app which subscribes to platform events and proxy the results to the client with Websockets. And when the user wants to unsub, he clicks on a button to send a request to the server that will then Unsubscribe and remove the user from the websocket. I will try again, because I remember I was using Node 12, maybe with 14 things change.

**Comment by AllanOricil** [comment]
I forgot to try using different node versions  😅 

I just remember I was creating a very simple vscode extension for platform events. Client was a webview, platform events subscriptions were managed by a node app, which would proxy responses to the client. I end up not publishing it because I couldn't make unsubscribe to work.
